### PR TITLE
scaffolder: add dry-run + Template Editor

### LINKED ISSUE #6568: Scaffolder: Have a way to do integration tests for fetch:template
The `fetch:template` action in the scaffolder doesn't support an out-of-action integration test path. With the cookiecutter templates, one could just run the CLI.

## Feature Suggestion

Offer the templating logic as an export of the scaffolder module, or even as CLI.

## Context

Authors of templates would like to run intergration tests, or also use their templates outside of the backstage creation pipeline.

### PR DESCRIPTION
**Pull request by Rugvip**
## Hey, I just made a Pull Request!

This adds a new `/v2/dry-run` endpoint to the scaffolder backend. It receives a serialized template, including local files, and returns a serialized result of running the template. The execution happens directly in the scaffolder backend and unlike regular scaffolder task jobs it's synchronous, without streaming of logs etc.

To support the dry-run endpoint, actions can now declare whether they support dry-run or not, with not supporting it being the default. An action that supports dry-run is responsible for checking `ctx.isDryRun` on the handler context and to behave accordingly.

In addition to the dry-run endpoint in the backend, this adds a new template editor to the frontend. It builds upon the existing form preview, but uses the web directory access API to load in templated from the local filesystem for editing. This lets users both edit the entire template as well as template contents, as well as trying it out in practice using the new dry-run capability. Each dry-run result is stored in the frontend, where you can inspect the directory contents, log, and outputs.

Fixes #6568

Still a bit experimental but opening for review. We should likely consider adding this behind some sort of flag.

https://user-images.githubusercontent.com/4984472/165973801-24bffaee-53f6-4573-849d-827c22a0850a.mov

#### :heavy_check_mark: Checklist

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Review by benjdlambert — COMMENTED** [review]
Let's get this in :ship:

**Review by benjdlambert — APPROVED** [review]

File: plugins/scaffolder-backend/src/scaffolder/actions/builtin/publish/githubPullRequest.ts [hunk 1]
```diff
@@ -219,56 +217,39 @@ export const createPublishGithubPullRequestAction = ({
         ? resolveSafeChildPath(ctx.workspacePath, sourcePath)
         : ctx.workspacePath;
 
-      const localFilePaths = await globby(['./**', './**/.*', '!.git'], {
-        cwd: fileRoot,
+      const directoryContents = await serializeDirectoryContents(fileRoot, {
```
Comment (benjdlambert) [inline plugins/scaffolder-backend/src/scaffolder/actions/builtin/publish/githubPullRequest.ts:220]: Do we need to filter out the `.git` and does this also include hidden files and folders?

**Comment by wrighbr** [comment]
@Rugvip I'm getting the below error message when trying use this function
`Failed to execute 'btoa' on 'Window': The string to be encoded contains characters outside of the Latin1 range.`

root cause: .git folder has non ascii characters in the files

**Comment by dotressel** [comment]
Hi @Rugvip , props for your work! This is a great feature.
I'm wondering if there is any documentation yet?

**Comment by Rugvip** [comment]
@dotressel Thank you! No docs beyond the pretty slim API surface. What kind of docs would you be interested in? For integrators and how to build out dry-running capabilities, or something more end-user oriented like where to find the template editor and how to use it?

**Comment by dotressel** [comment]
I'd be interested in using the dry-run in our CI to continuously test our templates. I don't know if that is even possible but if so, it would be great to see this documented :)

**Comment by sp71** [comment]
@Rugvip Great update! Would there be a way to interact with the generated template? The dry run logs shows a location for the generated output but it's always empty. I'm assuming it gets deleted on dry run. Any way we can not delete it? It would help immensely in debugging issues

**Pull request merged**